File: 649-dota2-senate/dota2-senate.py
from collections import deque

class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        n = len(senate)
        r_queue = deque(
            i for i, sen in enumerate(senate) if sen == 'R'
        )
        d_queue = deque(
            i for i, sen in enumerate(senate) if sen == 'D'
        )

        while r_queue and d_queue:
            r_turn, d_turn = r_queue.popleft(), d_queue.popleft()
            
            if r_turn < d_turn:
                r_queue.append(r_turn + n)
            else:
                d_queue.append(d_turn + n)
        
        return 'Radiant' if r_queue else 'Dire'

File: 649-dota2-senate/dota2-senate.java
class Solution {
    public String predictPartyVictory(String senate) {
        int len = senate.length();
        int [] r = new int[ len ] ,
               d = new int[ len ] ;
        int re = 0 , rd = 0 , de = 0 , dd = 0 ;
        for ( int i = 0 ; i < len ; i++ ) {
            if ( senate.charAt( i ) == 'R' ) r[rd++] = i ;
            else d[dd++] = i ;
        }
        while ( rd-re > 0 && dd-de > 0 ) {
            int i = r[re++] ;
            int j = d[de++] ;
            if ( i < j ) {
                r[(rd%len)] = j+len ;
                rd++ ;
            } else {
                d[(dd%len)] = i+len ;
                dd++ ;
            }
        }
        return ( rd-re > 0 ) ? "Radiant" : "Dire" ;
    }
}